File: packages/go/xlsx/xlsx_test.go
//go:build sheetcompressor_excelize

package xlsx

import (
	"bytes"
	"errors"
	"reflect"
	"testing"

	"github.com/xuri/excelize/v2"

	"github.com/mythopoeic/sheet-compressor/packages/go/sheetcompressor"
)

/* -------------------------------------------------------------------------- */
/* Test helpers — build small in-memory .xlsx files using excelize itself.    */
/*                                                                            */
/* The adapter is tested against bytes excelize produces (round-trip), so a   */
/* future excelize bump or behaviour change surfaces here loudly. Asserts     */
/* are on the Grid (rows, origin, cellMeta, charts) — NEVER on the            */
/* compressed output (see SPEC §8.3).                                         */
/* -------------------------------------------------------------------------- */

func writeWorkbook(t *testing.T, build func(f *excelize.File)) []byte {
	t.Helper()
	f := excelize.NewFile()
	t.Cleanup(func() { _ = f.Close() })
	build(f)
	var buf bytes.Buffer
	if err := f.Write(&buf); err != nil {
		t.Fatalf("excelize write: %v", err)
	}
	return buf.Bytes()
}

func mustRead(t *testing.T, data []byte, opts Options) *sheetcompressor.Grid {
	t.Helper()
	g, err := ReadSheet(data, opts)
	if err != nil {
		t.Fatalf("ReadSheet: %v", err)
	}
	return g
}

func dataTypes(meta [][]sheetcompressor.CellMeta) [][]sheetcompressor.DataType {
	out := make([][]sheetcompressor.DataType, len(meta))
	for i, row := range meta {
		out[i] = make([]sheetcompressor.DataType, len(row))
		for j, m := range row {
			out[i][j] = m.DataType
		}
	}
	return out
}

/* -------------------------------------------------------------------------- */
/* Empty / minimal workbooks                                                  */
/* -------------------------------------------------------------------------- */

func TestReadSheet_EmptyWorkbook(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {})
	g := mustRead(t, data, Options{})
	if len(g.Rows) != 0 {
		t.Errorf("expected empty Rows, got %#v", g.Rows)
	}
	if g.Origin != (sheetcompressor.Origin{Row: 1, Col: 1}) {
		t.Errorf("expected origin {1,1}, got %+v", g.Origin)
	}
	if g.CellMeta != nil {
		t.Errorf("expected nil CellMeta for empty sheet, got %#v", g.CellMeta)
	}
	if len(g.Charts) != 0 {
		t.Errorf("expected no charts, got %d", len(g.Charts))
	}
}

func TestReadSheet_A1AnchoredGrid(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "Name")
		f.SetCellValue("Sheet1", "B1", "Qty")
		f.SetCellValue("Sheet1", "A2", "Apple")
		f.SetCellValue("Sheet1", "B2", 3)
	})
	g := mustRead(t, data, Options{})
	if g.Origin != (sheetcompressor.Origin{Row: 1, Col: 1}) {
		t.Errorf("expected origin {1,1}, got %+v", g.Origin)
	}
	want := [][]string{
		{"Name", "Qty"},
		{"Apple", "3"},
	}
	if !reflect.DeepEqual(g.Rows, want) {
		t.Errorf("rows mismatch\n got: %#v\nwant: %#v", g.Rows, want)
	}
}

func TestReadSheet_RespectsTrueOrigin(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "C5", "Name")
		f.SetCellValue("Sheet1", "D5", "Qty")
		f.SetCellValue("Sheet1", "C6", "Apple")
		f.SetCellValue("Sheet1", "D6", 3)
	})
	g := mustRead(t, data, Options{})
	if g.Origin != (sheetcompressor.Origin{Row: 5, Col: 3}) {
		t.Errorf("expected origin {5,3}, got %+v", g.Origin)
	}
	want := [][]string{
		{"Name", "Qty"},
		{"Apple", "3"},
	}
	if !reflect.DeepEqual(g.Rows, want) {
		t.Errorf("rows mismatch\n got: %#v\nwant: %#v", g.Rows, want)
	}
}

func TestReadSheet_FillsInternalGapsAsEmpty(t *testing.T) {
	// Row 1 has only B1; row 2 has A2 and C2. Used range is A1:C2; gaps inside
	// must surface as "" so downstream stages see a rectangular grid.
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "B1", "B1val")
		f.SetCellValue("Sheet1", "A2", "A2val")
		f.SetCellValue("Sheet1", "C2", 7)
	})
	g := mustRead(t, data, Options{})
	want := [][]string{
		{"", "B1val", ""},
		{"A2val", "", "7"},
	}
	if !reflect.DeepEqual(g.Rows, want) {
		t.Errorf("rows mismatch\n got: %#v\nwant: %#v", g.Rows, want)
	}
	if g.Origin != (sheetcompressor.Origin{Row: 1, Col: 1}) {
		t.Errorf("expected origin {1,1}, got %+v", g.Origin)
	}
}

/* -------------------------------------------------------------------------- */
/* Sheet selection                                                            */
/* -------------------------------------------------------------------------- */

func TestReadSheet_DefaultsToFirstSheet(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "A")
		f.NewSheet("Second")
		f.SetCellValue("Second", "A1", "B")
	})
	g := mustRead(t, data, Options{})
	if got, want := g.Rows[0][0], "A"; got != want {
		t.Errorf("expected first-sheet value %q, got %q", want, got)
	}
}

func TestReadSheet_SelectByName(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "A")
		f.NewSheet("Second")
		f.SetCellValue("Second", "A1", "B")
	})
	g := mustRead(t, data, Options{SheetName: "Second"})
	if got, want := g.Rows[0][0], "B"; got != want {
		t.Errorf("expected second-sheet value %q, got %q", want, got)
	}
}

func TestReadSheet_SelectByIndex(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "A")
		f.NewSheet("Second")
		f.SetCellValue("Second", "A1", "B")
	})
	g := mustRead(t, data, Options{SheetIndex: 1})
	if got, want := g.Rows[0][0], "B"; got != want {
		t.Errorf("expected index-1 value %q, got %q", want, got)
	}
}

func TestReadSheet_UnknownSheetName(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "A")
	})
	if _, err := ReadSheet(data, Options{SheetName: "Missing"}); err == nil {
		t.Fatal("expected error for unknown sheet name, got nil")
	}
}

func TestReadSheet_IndexOutOfRange(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "A")
	})
	if _, err := ReadSheet(data, Options{SheetIndex: 5}); err == nil {
		t.Fatal("expected error for out-of-range index, got nil")
	}
}

/* -------------------------------------------------------------------------- */
/* cellMeta dataType                                                          */
/* -------------------------------------------------------------------------- */

func TestReadSheet_DataTypeMapping(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "hello") // text
		f.SetCellValue("Sheet1", "B1", 42)      // number
		f.SetCellBool("Sheet1", "C1", true)     // bool
		f.SetCellFormula("Sheet1", "D1", "=1+1")
	})
	g := mustRead(t, data, Options{})
	if g.CellMeta == nil {
		t.Fatal("expected cellMeta")
	}
	got := dataTypes(g.CellMeta)
	want := [][]sheetcompressor.DataType{
		{
			sheetcompressor.DataTypeText,
			sheetcompressor.DataTypeNumber,
			sheetcompressor.DataTypeBool,
			sheetcompressor.DataTypeFormula,
		},
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("dataType mismatch\n got: %#v\nwant: %#v", got, want)
	}
}

func TestReadSheet_GapCellsAreEmpty(t *testing.T) {
	// Used range A1:C1 with a gap at B1 → meta has dataType="empty" in the middle.
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "x")
		f.SetCellValue("Sheet1", "C1", 3)
	})
	g := mustRead(t, data, Options{})
	if g.CellMeta == nil {
		t.Fatal("expected cellMeta")
	}
	got := dataTypes(g.CellMeta)
	want := [][]sheetcompressor.DataType{
		{
			sheetcompressor.DataTypeText,
			sheetcompressor.DataTypeEmpty,
			sheetcompressor.DataTypeNumber,
		},
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("dataType mismatch\n got: %#v\nwant: %#v", got, want)
	}
}

func TestReadSheet_OmitsCellMetaForEmptySheet(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {})
	g := mustRead(t, data, Options{})
	if g.CellMeta != nil {
		t.Errorf("expected nil CellMeta on empty sheet, got %#v", g.CellMeta)
	}
}

/* -------------------------------------------------------------------------- */
/* Chart extraction                                                           */
/* -------------------------------------------------------------------------- */

func TestReadSheet_ExtractsBarChartWithTitleAxesSeries(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "Quarter")
		f.SetCellValue("Sheet1", "B1", "Sales")
		f.SetCellValue("Sheet1", "A2", "Q1")
		f.SetCellValue("Sheet1", "B2", 100)
		f.SetCellValue("Sheet1", "A3", "Q2")
		f.SetCellValue("Sheet1", "B3", 150)
		if err := f.AddChart("Sheet1", "B5", &excelize.Chart{
			Type:  excelize.Bar,
			Title: []excelize.RichTextRun{{Text: "Sales"}},
			Series: []excelize.ChartSeries{
				{
					Name:       "Sheet1!$B$1",
					Categories: "Sheet1!$A$2:$A$3",
					Values:     "Sheet1!$B$2:$B$3",
				},
			},
			XAxis: excelize.ChartAxis{Title: []excelize.RichTextRun{{Text: "Quarter"}}},
			YAxis: excelize.ChartAxis{Title: []excelize.RichTextRun{{Text: "Amount"}}},
		}); err != nil {
			t.Fatalf("AddChart: %v", err)
		}
	})
	g := mustRead(t, data, Options{})
	if len(g.Charts) != 1 {
		t.Fatalf("expected 1 chart, got %d", len(g.Charts))
	}
	c := g.Charts[0]
	if c.Type != "bar" {
		t.Errorf("expected type bar, got %s", c.Type)
	}
	if c.AnchorRange == "" {
		t.Errorf("expected non-empty anchorRange")
	}
	if c.Title == nil || *c.Title != "Sales" {
		t.Errorf("expected title \"Sales\", got %v", c.Title)
	}
	if c.Axes == nil || c.Axes.X == nil || *c.Axes.X != "Quarter" {
		t.Errorf("expected x axis \"Quarter\", got %v", c.Axes)
	}
	if c.Axes == nil || c.Axes.Y == nil || *c.Axes.Y != "Amount" {
		t.Errorf("expected y axis \"Amount\", got %v", c.Axes)
	}
	if !reflect.DeepEqual(c.Series, []string{"B1"}) {
		t.Errorf("expected series [B1], got %#v", c.Series)
	}
	if !reflect.DeepEqual(c.DataRanges, []string{"B2:B3"}) {
		t.Errorf("expected dataRanges [B2:B3], got %#v", c.DataRanges)
	}
}

func TestReadSheet_ChartTypeMapping(t *testing.T) {
	cases := []struct {
		ct   excelize.ChartType
		want sheetcompressor.ChartType
	}{
		{excelize.Bar, "bar"},
		{excelize.Line, "line"},
		{excelize.Pie, "pie"},
		{excelize.Scatter, "scatter"},
		{excelize.Area, "area"},
	}
	for _, tc := range cases {
		tc := tc
		t.Run(string(tc.want), func(t *testing.T) {
			data := writeWorkbook(t, func(f *excelize.File) {
				f.SetCellValue("Sheet1", "A1", "x")
				f.SetCellValue("Sheet1", "B1", 1)
				if err := f.AddChart("Sheet1", "B5", &excelize.Chart{
					Type: tc.ct,
					Series: []excelize.ChartSeries{
						{Name: "Sheet1!$A$1", Values: "Sheet1!$B$1:$B$1"},
					},
				}); err != nil {
					t.Fatalf("AddChart: %v", err)
				}
			})
			g := mustRead(t, data, Options{})
			if len(g.Charts) != 1 {
				t.Fatalf("expected 1 chart, got %d", len(g.Charts))
			}
			if g.Charts[0].Type != tc.want {
				t.Errorf("type mismatch: got %s want %s", g.Charts[0].Type, tc.want)
			}
		})
	}
}

func TestReadSheet_NoChartsReturnsEmpty(t *testing.T) {
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "x")
	})
	g := mustRead(t, data, Options{})
	if len(g.Charts) != 0 {
		t.Errorf("expected no charts, got %d", len(g.Charts))
	}
}

/* -------------------------------------------------------------------------- */
/* Error paths                                                                */
/* -------------------------------------------------------------------------- */

func TestReadSheet_RejectsGarbageBytes(t *testing.T) {
	if _, err := ReadSheet([]byte("not a workbook"), Options{}); err == nil {
		t.Fatal("expected error for garbage input")
	}
}

func TestReadSheet_ReadSheetFile_Missing(t *testing.T) {
	if _, err := ReadSheetFile("/tmp/does-not-exist.xlsx", Options{}); err == nil {
		t.Fatal("expected error for missing file")
	}
}

func TestReadSheet_ErrAdapterUnavailable_NotReturnedInRealBuild(t *testing.T) {
	// Sanity check: the real adapter must NEVER return ErrAdapterUnavailable.
	// (The stub does; pinning here makes the build-tag wiring explicit.)
	data := writeWorkbook(t, func(f *excelize.File) {
		f.SetCellValue("Sheet1", "A1", "x")
	})
	_, err := ReadSheet(data, Options{})
	if errors.Is(err, ErrAdapterUnavailable) {
		t.Fatal("real adapter returned ErrAdapterUnavailable")
	}
}

File: packages/typescript/test/adapters/xlsx.test.ts
import { describe, expect, it } from "vitest";
import * as XLSX from "xlsx";

import { readSheet } from "../../src/adapters/xlsx.ts";
import type { Grid } from "../../src/index.ts";

import { buildChartXlsx } from "./xlsxBuilder.ts";

/** Build an .xlsx in-memory from a 2D array of cells using SheetJS. */
function buildBuffer(
  data: (string | number | boolean | Date | null)[][],
  opts: { sheetName?: string; origin?: string } = {},
): Buffer {
  const wb = XLSX.utils.book_new();
  const ws = XLSX.utils.aoa_to_sheet(data, { cellDates: true });
  if (opts.origin) {
    // aoa_to_sheet always anchors at A1; rewrite ref + shift cells if needed.
    const range = XLSX.utils.decode_range(ws["!ref"] as string);
    const origin = XLSX.utils.decode_cell(opts.origin);
    const dr = origin.r - range.s.r;
    const dc = origin.c - range.s.c;
    if (dr !== 0 || dc !== 0) {
      const moved: Record<string, unknown> = {};
      for (const k of Object.keys(ws)) {
        if (k.startsWith("!")) continue;
        const c = XLSX.utils.decode_cell(k);
        const nc = { r: c.r + dr, c: c.c + dc };
        moved[XLSX.utils.encode_cell(nc)] = ws[k];
      }
      for (const k of Object.keys(ws)) if (!k.startsWith("!")) delete ws[k];
      Object.assign(ws, moved);
      ws["!ref"] = XLSX.utils.encode_range({
        s: { r: range.s.r + dr, c: range.s.c + dc },
        e: { r: range.e.r + dr, c: range.e.c + dc },
      });
    }
  }
  XLSX.utils.book_append_sheet(wb, ws, opts.sheetName ?? "Sheet1");
  return XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
}

describe("readSheet — empty / minimal workbooks", () => {
  it("returns an empty grid for a workbook with no cells", () => {
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(wb, {}, "Sheet1");
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    const g: Grid = readSheet(buf);
    expect(g.rows).toEqual([]);
    expect(g.origin).toEqual({ row: 1, col: 1 });
    expect(g.charts ?? []).toEqual([]);
  });

  it("returns rows + origin {1,1} for an A1-anchored 2x2 grid", () => {
    const buf = buildBuffer([
      ["Name", "Qty"],
      ["Apple", 3],
    ]);
    const g = readSheet(buf);
    expect(g.rows).toEqual([
      ["Name", "Qty"],
      ["Apple", "3"],
    ]);
    expect(g.origin).toEqual({ row: 1, col: 1 });
  });

  it("respects the sheet's true origin when data starts away from A1", () => {
    const buf = buildBuffer(
      [
        ["Name", "Qty"],
        ["Apple", 3],
      ],
      { origin: "C5" },
    );
    const g = readSheet(buf);
    expect(g.origin).toEqual({ row: 5, col: 3 });
    expect(g.rows).toEqual([
      ["Name", "Qty"],
      ["Apple", "3"],
    ]);
  });

  it("fills cell-internal gaps with empty strings (no ragged rows)", () => {
    // Row 1 has B1 only, row 2 has A2 and C2 — range spans A1:C2.
    const wb = XLSX.utils.book_new();
    const ws: Record<string, unknown> = {
      "!ref": "A1:C2",
      B1: { t: "s", v: "B1val" },
      A2: { t: "s", v: "A2val" },
      C2: { t: "n", v: 7 },
    };
    XLSX.utils.book_append_sheet(wb, ws, "Sheet1");
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    const g = readSheet(buf);
    expect(g.rows).toEqual([
      ["", "B1val", ""],
      ["A2val", "", "7"],
    ]);
  });
});

describe("readSheet — sheet selection", () => {
  it("defaults to the first sheet", () => {
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["A"]]),
      "First",
    );
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["B"]]),
      "Second",
    );
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    expect(readSheet(buf).rows).toEqual([["A"]]);
  });

  it("selects a sheet by name", () => {
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["A"]]),
      "First",
    );
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["B"]]),
      "Second",
    );
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    expect(readSheet(buf, { sheet: "Second" }).rows).toEqual([["B"]]);
  });

  it("selects a sheet by 0-indexed position", () => {
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["A"]]),
      "First",
    );
    XLSX.utils.book_append_sheet(
      wb,
      XLSX.utils.aoa_to_sheet([["B"]]),
      "Second",
    );
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    expect(readSheet(buf, { sheet: 1 }).rows).toEqual([["B"]]);
  });
});

describe("readSheet — cellMeta dataType", () => {
  it("maps SheetJS cell types to the core's DataType vocabulary", () => {
    const wb = XLSX.utils.book_new();
    const ws: Record<string, unknown> = {
      "!ref": "A1:F1",
      A1: { t: "s", v: "hello" },
      B1: { t: "n", v: 42 },
      C1: { t: "b", v: true },
      D1: { t: "d", v: new Date("2024-01-15T00:00:00Z") },
      E1: { t: "n", v: 7, f: "1+6" },
      F1: { t: "e", v: 7 },
    };
    XLSX.utils.book_append_sheet(wb, ws, "Sheet1");
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    const g = readSheet(buf);
    expect(g.cellMeta).toBeDefined();
    const types = g.cellMeta![0]!.map((m) => m.dataType);
    expect(types).toEqual([
      "text",
      "number",
      "bool",
      "date",
      "formula",
      "error",
    ]);
  });

  it("marks gap cells inside the used range as empty in cellMeta", () => {
    const wb = XLSX.utils.book_new();
    const ws: Record<string, unknown> = {
      "!ref": "A1:C1",
      A1: { t: "s", v: "x" },
      C1: { t: "n", v: 3 },
    };
    XLSX.utils.book_append_sheet(wb, ws, "Sheet1");
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    const g = readSheet(buf);
    expect(g.cellMeta).toBeDefined();
    expect(g.cellMeta![0]!.map((m) => m.dataType)).toEqual([
      "text",
      "empty",
      "number",
    ]);
  });

  it("omits cellMeta entirely when the workbook has no cell type info", () => {
    // Empty workbook — no cells at all → no metadata to emit.
    const wb = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(wb, {}, "Sheet1");
    const buf = XLSX.write(wb, { type: "buffer", bookType: "xlsx" }) as Buffer;
    const g = readSheet(buf);
    expect(g.cellMeta).toBeUndefined();
  });
});

describe("readSheet — chart extraction", () => {
  it("extracts a single embedded bar chart with title, axes, series, data", () => {
    const buf = buildChartXlsx({
      data: [
        ["Quarter", "Sales"],
        ["Q1", 100],
        ["Q2", 150],
        ["Q3", 200],
      ],
      chart: {
        chartType: "bar",
        // anchor from B5 to F20 (xdr uses 0-indexed)
        anchor: { fromCol: 1, fromRow: 4, toCol: 5, toRow: 19 },
        name: "Q1Sales",
        title: "Sales",
        xAxisTitle: "Quarter",
        yAxisTitle: "Amount",
        series: [
          {
            nameLiteral: "Sales",
            valuesRange: "Sheet1!$B$2:$B$4",
          },
        ],
      },
    });
    const g = readSheet(buf);
    expect(g.charts).toBeDefined();
    expect(g.charts).toHaveLength(1);
    const c = g.charts![0]!;
    expect(c.name).toBe("Q1Sales");
    expect(c.type).toBe("bar");
    expect(c.anchorRange).toBe("B5:F20");
    expect(c.title).toBe("Sales");
    expect(c.axes).toEqual({ x: "Quarter", y: "Amount" });
    expect(c.series).toEqual(["Sales"]);
    expect(c.dataRanges).toEqual(["B2:B4"]);
  });

  it("maps SheetJS chart-type elements to the ChartType vocabulary", () => {
    for (const t of ["bar", "line", "pie", "scatter", "area"] as const) {
      const buf = buildChartXlsx({
        data: [["x"]],
        chart: {
          chartType: t,
          anchor: { fromCol: 0, fromRow: 0, toCol: 1, toRow: 1 },
          name: t,
        },
      });
      expect(readSheet(buf).charts![0]!.type).toBe(t);
    }
  });

  it("returns charts: [] when the workbook has no charts", () => {
    const buf = buildBuffer([["x"]]);
    expect(readSheet(buf).charts ?? []).toEqual([]);
  });
});
